src/Test/Internal.elm

```diff
@@ -53,40 +58,55 @@ fuzzTest fuzzer desc getExpectation =
        Whether it passes or fails, do this n times
     -}
     let
-        getFailures failures currentSeed remainingRuns =
+        getFailures failures currentSeed remainingRuns results =
             let
                 genVal =
                     unpackGenVal fuzzer
 
                 ( value, nextSeed ) =
                     Random.step genVal currentSeed
-
-                newFailures =
-                    case getExpectation value of
-                        Pass ->
-                            failures
-
-                        failedExpectation ->
-                            let
-                                genTree =
-                                    unpackGenTree fuzzer
-
-                                ( rosetree, nextSeedAgain ) =
-                                    Random.step genTree currentSeed
-                            in
-                                shrinkAndAdd rosetree getExpectation failedExpectation failures
             in
-                if remainingRuns == 1 then
-                    newFailures
-                else
-                    getFailures newFailures nextSeed (remainingRuns - 1)
+                case Dict.get (toString value) results of
+                    Just _ ->
+                        -- we can skip this, already have the result in `failures`
+                        failures
+
+                    Nothing ->
```

Comment:
I suspect we should be extracting some functions here, and possibly even record types, but that's out of scope of this PR.